### PR TITLE
curvefs: implement setxattr interface.

### PR DESCRIPTION
**Pull request by SeanHai**
Signed-off-by: wanghai01 <seanhaizi@163.com>

### What problem does this PR solve?

Issue Number: #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/client/curve_fuse_op.cpp [hunk 1]
```diff
@@ -322,17 +322,22 @@ int Warmup(const std::string& name, const std::string& value) {
 
 void FuseOpSetXattr(fuse_req_t req, fuse_ino_t ino, const char* name,
                     const char* value, size_t size, int flags) {
-    // only support "curvefs.warmup.op" xattr
     std::string xattrValue(value, size);
     VLOG(9) << "FuseOpSetXattr"
             << " ino " << ino << " name " << name << " value " << xattrValue
             << " flags " << flags;
-    int code = ENOTSUP;
+    int code = 0;
```
Comment (cw123) [inline curvefs/src/client/curve_fuse_op.cpp:329]: code only use in if, move it in the if

File: curvefs/src/client/fuse_client.cpp [hunk 2]
```diff
@@ -1104,46 +1104,45 @@ CURVEFS_ERROR FuseClient::FuseOpGetXattr(fuse_req_t req, fuse_ino_t ino,
...
+
+    auto it = inodeAttr.xattr().find(name);
+    if (it != inodeAttr.xattr().end()) {
+        xValue = it->second;
     }
 
-    CURVEFS_ERROR ret = CURVEFS_ERROR::NODATA;
+    ret = CURVEFS_ERROR::NODATA;
     if (xValue.length() > 0) {
         if ((size == 0 && xValue.length() <= MAXXATTRLENGTH) ||
```
Comment (cw123) [inline curvefs/src/client/fuse_client.cpp:1147]: what does size == 0 mean？
size == 0是什么意思，这时候不会出错吗？

File: curvefs/src/client/curve_fuse_op.cpp (hunk 1 above)
Comment (SeanHai) [reply curvefs/src/client/curve_fuse_op.cpp:329]: fixed

File: curvefs/src/client/fuse_client.cpp (hunk 2 above)
Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp:1147]: /**
 * Get an extended attribute
 *
 * If size is zero, the size of the value should be sent with
 * fuse_reply_xattr.
 *
 * If the size is non-zero, and the value fits in the buffer, the
 * value should be sent with fuse_reply_buf.
 *
 * If the size is too small for the value, the ERANGE error should
 * be sent.

size ==0 means you should return value's length by fuse_reply_xattr, and this flag used to get the value's length first by kernel and the real size will be set in the next request.

Comment (cw123) [reply curvefs/src/client/fuse_client.cpp:1147]: if size > MAXXATTRLENGTH && xValue.length() <= MAXXATTRLENGTH
it will return out of range?

Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp:1147]: fixed

**Review by cw123 — APPROVED** [review]

File: curvefs/src/client/fuse_client.cpp [hunk 3]
```diff
@@ -1156,6 +1155,41 @@ CURVEFS_ERROR FuseClient::FuseOpGetXattr(fuse_req_t req, fuse_ino_t ino,
     return ret;
 }
 
+CURVEFS_ERROR FuseClient::FuseOpSetXattr(fuse_req_t req, fuse_ino_t ino,
+                                         const char* name, const char* value,
+                                         size_t size, int flags) {
+    if (option_.disableXattr) {
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_client.cpp:1161]: VLOG(1) << "FuseOpSetXattr " << ****;

File: curvefs/src/client/fuse_client.cpp [hunk 4]
```diff
@@ -1104,49 +1104,48 @@ CURVEFS_ERROR FuseClient::FuseOpGetXattr(fuse_req_t req, fuse_ino_t ino,
...
-        if (inodeAttr.type() == FsFileType::TYPE_DIRECTORY) {
-            // not enable record summary info in dir xattr,
-            // need recursive computation all files
-            if (!enableSumInDir_) {
-                if (IsOneLayer(name)) {
-                    ret = xattrManager_->CalOneLayerSumInfo(&inodeAttr);
-                } else {
-                    ret = xattrManager_->CalAllLayerSumInfo(&inodeAttr);
-                }
+    // get summary info
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_client.cpp:1115]: The following conditions are different, please describe the function of this code as a whole.

File: curvefs/src/client/fuse_client.cpp [hunk 5]
```diff
@@ -1156,6 +1155,41 @@ CURVEFS_ERROR FuseClient::FuseOpGetXattr(fuse_req_t req, fuse_ino_t ino,
...
+    CURVEFS_ERROR ret = inodeManager_->GetInode(ino, inodeWrapper);
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "inodeManager get inode fail, ret = " << ret
+                   << ", inodeid = " << ino;
+        return ret;
+    }
+
+    ::curve::common::UniqueLock lgGuard = inodeWrapper->GetUniqueLock();
+    inodeWrapper->SetXattrLocked(strname, strvalue);
+    ret = inodeWrapper->SyncAttr();
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_client.cpp:1184]: Why do you need to do sync every time here?

File: curvefs/src/client/fuse_client.cpp (hunk 3 above)
Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp:1161]: fixed

File: curvefs/src/client/fuse_client.cpp (hunk 4 above)
Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp:1115]: done

File: curvefs/src/client/fuse_client.cpp (hunk 5 above)
Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp:1184]: The xattr will lost when cache crashed If not update the xattr to metaserver

**Comment by SeanHai** [comment]
recheck

**Comment by SeanHai** [comment]
recheck

**Comment by SeanHai** [comment]
recheck

**Comment by SeanHai** [comment]
recheck

File: curvefs/src/client/fuse_client.cpp [hunk 6]
```diff
@@ -1104,49 +1104,49 @@ CURVEFS_ERROR FuseClient::FuseOpGetXattr(fuse_req_t req, fuse_ino_t ino,
...
-                    ret = xattrManager_->CalAllLayerSumInfo(&inodeAttr);
-                }
+    // get summary info if the xattr name is summary type
+    if (IsSummaryInfo(name) && inodeAttr.type() == FsFileType::TYPE_DIRECTORY) {
+        // if not enable record summary info in dir xattr,
+        // need recursive computation all files;
+        // otherwise only recursive computation all dirs.
+        if (!enableSumInDir_) {
+            if (IsOneLayer(name)) {
+                ret = xattrManager_->CalOneLayerSumInfo(&inodeAttr);
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_client.cpp:1122]: Maybe enableSumInDir_ is more suitable as a member of xattrManager_

File: curvefs/src/client/fuse_client.cpp [hunk 7]
```diff
@@ -1156,6 +1156,43 @@ CURVEFS_ERROR FuseClient::FuseOpGetXattr(fuse_req_t req, fuse_ino_t ino,
...
+    }
+
+    ::curve::common::UniqueLock lgGuard = inodeWrapper->GetUniqueLock();
+    inodeWrapper->SetXattrLocked(strname, strvalue);
+    ret = inodeWrapper->SyncAttr();
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "set xattr fail, ret = " << ret << ", inodeid = " << ino
+                   << ", name = " << strname << ", value = " << strvalue;
+        return ret;
+    }
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_client.cpp:1192]: VLOG(1) << "FuseOpSetXattr end" << ****

File: curvefs/src/client/fuse_client.cpp (hunk 6 above)
Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp]: fixed

File: curvefs/src/client/fuse_client.cpp (hunk 7 above)
Comment (SeanHai) [reply curvefs/src/client/fuse_client.cpp:1148]: fixed

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**